### PR TITLE
API implementation part 3

### PR DESCRIPTION
**Pull request by rkuris**
Probably easier to review commit-by-commit

### CODE REVIEW

File: firewood/Cargo.toml
Comment (richardpringle) [inline firewood/Cargo.toml:35]: 🚀

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:394]: 💩

Comment (richardpringle) [reply firewood/src/db.rs:394]: Why `block_in_place` over `spawn_blocking`. The latter is usually used when one is unsure of the outer context in which the blocking call is made. In other words, this is only faster if there aren't a bunch of other tasks running on the same thread. 

My intuition tells me that performance would be much more predictable using `spawn_blocking` but *could be* faster using `block_in_place`. I'm not necessarily opposed to it, but I think it should be justified.

File: firewood/src/v2/api.rs
Comment (richardpringle) [inline firewood/src/v2/api.rs:99]: 🚀

File: firewood/src/v2/api.rs
Comment (richardpringle) [inline firewood/src/v2/api.rs:183]: Need to update the doc-comments

File: firewood/src/v2/propose.rs
Comment (richardpringle) [inline firewood/src/v2/propose.rs:153]: Wasn't this supposed to return the latest view upon success? Now we have to go back to the database and get the latest view?

File: firewood/tests/v2api.rs
Comment (richardpringle) [inline firewood/tests/v2api.rs:1]: I think you should just rename this file to API. `v2` is just a placeholder for the code, right? We definitely aren't going to write any v1 tests at this point.

File: firewood/tests/v2api.rs
Comment (richardpringle) [inline firewood/tests/v2api.rs:28]: Do we not have a unit test somewhere for this?

File: firewood/tests/v2api.rs
Comment (richardpringle) [inline firewood/tests/v2api.rs:47]: Should only define the key and value once then use that definition everywhere. 

```suggestion
    let key = b"smoke";
    let value = b"test";

    // insert a single key/value
    let batch_put = BatchOp::Put {
        key,
        value,
    };
    let proposal: Arc<firewood::db::Proposal> = db.propose(vec![batch_put]).await?.into();
    proposal.commit().await?;

    // ensure the latest hash is different
    let latest = db.root_hash().await?;
    assert_ne!(empty_hash, latest);

    // fetch the view of the latest
    let view = db.revision(latest).await.unwrap();

    // check that the key/value is there
    assert_eq!(view.val(key).await.unwrap().unwrap(), value);
```
you can compare `Vec<T>` with `&[T]`.

File: firewood/tests/v2api.rs
Comment (richardpringle) [inline firewood/tests/v2api.rs:8]: The test would be simpler if you didn't use the test macro and simply
```rust
fn smoke() -> Result<(), Box<dyn Error>> {
    let cfg = // ...
    let tmpdbpath = tmp_dir().join("testdb");
    let db = PersistedDb::new(tmpdbpath, &cfg).unwrap();
    
    tokio::block_on(async move {
        // then put your test logic here. 
    });
}
```

Then `spawn_blocking` wouldn't be necessary

File: firewood/src/db.rs
Comment (richardpringle) [inline firewood/src/db.rs:395]:
```suggestion
        block_in_place(|| {
             self
                  .get_revision(&TrieHash(root_hash)))
                  .map(Arc::new)
                  .ok_or(api::Error::HashNotFound { provided: root_hash }
        });
```
That's my best guess at formatting

File: firewood/src/db.rs
Comment (richardpringle) [reply firewood/src/db.rs:394]: Please respond before next round of review

File: firewood/src/db.rs
Comment (richardpringle) [reply firewood/src/db.rs:395]: nit ^

File: firewood/src/v2/api.rs
Comment (richardpringle) [reply firewood/src/v2/api.rs:183]: Please make this change before next review

File: firewood/src/v2/propose.rs
Comment (richardpringle) [reply firewood/src/v2/propose.rs:153]: simply a question, should not block merge

File: firewood/tests/v2api.rs
Comment (richardpringle) [reply firewood/tests/v2api.rs:1]: Please do this... or `v2_api.rs` or `api_v2.rs`... but I think it should be removed.

File: firewood/tests/v2api.rs
Comment (richardpringle) [reply firewood/tests/v2api.rs:8]: Not important, just a suggestion

File: firewood/tests/v2api.rs
Comment (richardpringle) [reply firewood/tests/v2api.rs:28]: I think this belongs elsewhere, but shouldn't hold up PR

File: firewood/tests/v2api.rs
Comment (richardpringle) [reply firewood/tests/v2api.rs:47]: Please do this before merge

**Review by richardpringle — CHANGES_REQUESTED** [review]

File: firewood/src/db.rs
Comment (rkuris) [reply firewood/src/db.rs:394]: The trouble here is that I can't move `self` into the closure, and `block_in_place` can borrow self since the thread is guaranteed to be done with it before returning.

I think if this is slower or faster is in the eye of the beholder (or of the benchmark tool). Since we don't know, and there are lifetime issues, this seems like a decent choice.

File: firewood/src/v2/propose.rs
Comment (rkuris) [reply firewood/src/v2/propose.rs:153]: The issue here is that T is specified by the caller. This really needs T specified by the implementation. It's difficult to implement it to return the specific type asked for by the caller, so that's why I got rid of T completely.

This can be done, but I don't think it's critical so I'm just removing the generic here as a start (you'd have to do that anyway).

File: firewood/tests/v2api.rs
Comment (rkuris) [reply firewood/tests/v2api.rs:1]: Can do this as a followup, especially once v2 is the new standard.

File: firewood/tests/v2api.rs
Comment (rkuris) [reply firewood/tests/v2api.rs:28]: This is an integration/smoke test for v2::api::root_hash, which isn't really tested elsewhere right now.

Comment (richardpringle) [reply firewood/tests/v2api.rs:31]: It should be! We.can remove this once it is, though.

File: firewood/tests/v2api.rs
Comment (richardpringle) [inline firewood/tests/v2api.rs:34]: Nice

**Review by richardpringle — APPROVED** [review]

**Pull request merged**
